### ISSUE TITLE
Params, BN, Learned Distribution and FC width layer about FBnet and your implementation

### ISSUE THREAD

**Issue by zhaohui-yang** [JunrQ/NAS#19]
Thank you for sharing this great code!

I wonder if you have tested the released architecture of FBNETA,B and C? I calculated the parameters of FBNET-B, but it doesn't match with the origin paper.

| in_channels | out_channels | kernel | expansion | groups | stride | input | params |
| -- | -- | -- | -- | -- | -- | -- | -- | 
| 3 | 16 | 3 | 1 | 1 | 2 | 224 | 432 |
| 16 | 16 | 3 | 1 | 1 | 1 | 112 | 656 |
| 16 | 24 | 3 | 6 | 1 | 2 | 112 | 4704 |
|24 | 24 | 5 | 1 | 1 | 1 | 56 | 1752 |
| 24 | 24 | 3 | 1 | 1 | 1 | 56 | 1368 |
| 24 | 24 | 3 | 1 | 1 | 1 | 56 | 1368 |
| 24 | 32 | 5 | 6 | 1 | 2 | 56 | 11664 |
| 32 | 32 | 5 | 3 | 1 | 1 | 28 | 8544 |
| 32 | 32 | 3 | 6 | 1 | 1 | 28 | 14016 |
| 32 | 32 | 5 | 6 | 1 | 1 | 28 | 17088 |
|32 | 64 | 5 | 6 | 1 | 2 | 28 | 23232 |
| 64 | 64 | 5 | 1 | 1 | 1 | 14 | 9792 |
| 64 | 64 | 0 | 0 | 0 | 0 | 14 | 0 |
| 64 | 64 | 5 | 3 | 1 | 1 | 14 | 29376 |
| 64 | 112 | 5 | 6 | 1 | 1 | 14 | 77184 |
| 112 | 112 | 3 | 1 | 1 | 1 | 14 | 26096|
| 112 | 112 | 5 | 1 | 1 | 1 | 14 | 27888 |
|112 | 112 | 5 | 3 | 1 | 1 | 14 | 83664 |
| 112 | 184 | 5 | 6 | 1 | 2 | 14 | 215712 |
|184 | 184 | 5 | 1 | 1 | 1 | 7 | 72312 |
| 184 | 184 | 5 | 6 | 1 | 1 | 7 | 433872 |
| 184 | 184 | 5 | 6 | 1 | 1 | 7 | 433872 |
| 184 | 352 | 3 | 6 | 1 | 1 | 7 | 601680 |
| 352 | 1504 | 1 | 1 | 1 | 1 | 7 | 529408|
| 1504 | 1000 | 1 | 1 | 1 | 1 | 1 | 1504000 |
| | | | | | | | 4129680 |

The calculated parameters is 4.1M, which is lower than the reported 4.5M in the paper. The flops is also not consist. I wonder if you have calculated the number of parameters or flops?

Besides:

1. Did you test BN layer in implementation? From the paper and your implementation, I didn't find BN in the middle, however, it is used in Shufflenet v2. I wonder if you have tested the effect? will the performance be better is you remove the BN layer in the middle?

2. Have you ever searched architectures with latency on GPU. I first evaluate the lantency on my Titan XP GPU and search FBNet architectures and found it tends to select the most parameter module with almost no variance. However, in the original paper, the author says that he sampled 6 different architectures after training. I wonder if your searched architecture on GPU/CPU latency has the observation that the learned distribution could generate both large model like FBNET-C and light-weight model like FBNET-A

3. Why do you use 1984 as FC layer width? Tab. 1 in paper have 1504 and 1984, which make me confusing.

**Comment by JunrQ** [comment]
Hi @zhaohui-yang , thank you for your question.

For question 2, i think you can use the loss function as the original paper and try again. (the default in my code is different, sorry i didn't make it clear)

For 3, i think it might be a mistake... 

Sorry that i could't answer all your questions for the fact that i didn't do those researches.

**Comment by zhaohui-yang** [comment]
By the way, do you have the plan to rerun FBNET-B on ImageNet? And it is reimplemented and by [Single-Path NAS: Designing Hardware-Efficient ConvNets in less than 4 Hours](https://arxiv.org/pdf/1904.02877.pdf) and [Single Path One-Shot Neural Architecture Search with Uniform Sampling
](https://arxiv.org/pdf/1904.00420.pdf)

**Comment by JunrQ** [comment]
No... due to the job changes

**Comment by zhaohui-yang** [comment]
Thanks a lot~

**Comment by zhaohui-yang** [comment]
hi，

For Q1, not solved for FBNET-B

For Q2, I will evaluate in the near future.

For Q3, I think 1504 is enough(not calculate in detail), for most of works with mobilenet v2 module use 1280 for fc layer. Maybe it is enough. 

Thanks a lot~
